### ISSUE TITLE
Support clean option for failed import cluster in tendrl

### ISSUE THREAD

**Issue by shtripat** [Tendrl/commons#809]
If while import cluster at certain stage import fails, there should be an option available in tendrl to clean the partial details available for the cluster in tendrl, so that after correcting the issues user can import the cluster again.

@r0h4n @nthomas-redhat @julienlim @mbukatov add your thoughts comments please.

**Comment by julienlim** [comment]
@shtripat @r0h4n @nthomas-redhat @mbukatov @Tendrl/tendrl-qe @Tendrl/tendrl-core @gnehapk @cloudbehl @a2batic 

Is it possible that when a user triggers an "Import Cluster" that the Tendrl system detects that there are partial details already about the cluster in Tendrl and prompts user to confirm deletion / cleanup of the partial details (or override details after the import) before re-attempting to import the cluster?  

It's almost like using a "-f" (force) switch on a CLI command when the "import" is issued.

Thoughts?

**Comment by shtripat** [comment]
@julienlim need to try this. At first glance looks like it should be possible because based on different failures options while import we need to check what details could be present for the cluster and if cluster details available from those options, we need to ask user to force import (which internally would execute a cleanup first and then import again).

@r0h4n thoughts?

**Comment by r0h4n** [comment]
This shouldnt be another call altogether

In case of failures or retries, a combination of "unmanage cluster"   then "import cluster" should work

**Comment by r0h4n** [comment]
The "-f" (force) switch can indeed call "unmanage" before calling import

**Comment by julienlim** [comment]
@shtripat @r0h4n Sounds like we're leaning towards "-f" (force) as the approach for dealing with failures or retries.  

From a UI perspective, if we don't need to ask user to do more (i.e. confirming clearing up the cluster details or unmanaged), that would be preferable - so in a sense from a UI perspective, it's done automatically for the user.

**Comment by r0h4n** [comment]
Yes, can you update the UX designs for this? We can put this task in once umanage is done

**Comment by julienlim** [comment]
@r0h4n  @Tendrl/tendrl-core @Tendrl/tendrl-qe @mcarrano @gnehapk @a2batic @cloudbehl 

I don't believe there's additional changes needed on the Import side, but perhaps on the Task side.  Specifically, the UI workflow is this:

(1) Using Tendrl UI, in [All Cluster (list)](https://redhat.invisionapp.com/share/8QCOEVEY9#/screens/244738628), user selects a cluster to import by clicking on "Import" button.

(2) User is presented the ["Import Cluster" options](https://redhat.invisionapp.com/share/8QCOEVEY9#/screens/247445416) and clicks on "Import" to proceed with the import.

Since Tendrl is automatically doing the "unmanage" as part of the Import flow, as mentioned previously, we don't need to ask user to choose or acknowledge more options in the UI.  

(3) Tendrl UI tells user the "Import Cluster Task" is submitted, and provides user a way to see the Import Cluster details.

Right now, I am assuming this "Import Cluster transaction" -- comprised of unmanage cluster and import cluster -- will generate 2 flows, i.e. unmanage cluster and then import cluster.  So, when user goes to see the task details for the Import Cluster, can these 2 flows show up as a single task in the Task Details and Task List view?

**Comment by a2batic** [comment]
@julienlim @Tendrl/tendrl-core @shtripat @gnehapk @cloudbehl 

Following is what I understood from UI perspective:-

**The cleanup option is a combination of unmanage + import**
When import fails for a cluster, user will be asked to confirm deletion/cleanup of partial details and re-import will be done automatically after confirmation.

I have following doubts:-
1. Unmanage action will trigger only when deletion/cleanup done, or it will be trigger every time import happens ?
2. How the cleanup option will be shown? Through a dialog box, just like unmanage [1], or it will appear as a button next to import button (import button will be disabled because import fails) or the import button will be replaced with something like deletion/cleanup or re-import?

**Comment by julienlim** [comment]
@a2batic  @Tendrl/tendrl-core @Tendrl/tendrl-qe  @shtripat @gnehapk @cloudbehl @mcarrano @jjkabrown1

Based on Architecture discussions yesterday (See [Architecture Notes for 6 Feb 2018]((https://github.com/Tendrl/documentation/wiki/Architecture-Meeting-Notes)), we have decided that user does have to perform Unmanage first before doing an Import, so they are 2 discrete operations after all and not combined.

So what that means:
(1) Unmanage action is always visible.  This way if cleanup can occur before an Import is performed, or after a failed Import, or successful import.  
* A nice-to-have to allow for a better user experience would be for the Unmanage to appear the very first time a cluster is detected and has not yet been imported.

(2) Unmanage, as well as Enable/Disable Profiling actions [buttons] cannot be performed while an Import [Cluster] is in progress, i.e. it becomes disabled during Cluster Import execution.

(3) For Import Cluster in the UI, I still think it's desirable and provides for a "smarter" user experience for the clean-up to be automatically handled when Cluster is imported (if there is existing data in etcd for the cluster OR if the cluster import failed previously), rather than having to prompt user to unmanage to do it.  From a backend perspective, they are definitely separate and discrete flows (and that makes sense), and with a UI, you want to make it intuitive and easy for the user.
* If we don't do it automatically for the user from a UI perspective, what this means is that we do one of the following:
(3.1) have to inform the user that he/she needs to perform an "Unmanage" prior to performing the "Import" [Cluster] if we can detect the existence of etcd data or failed import cluster task/job previously
(3.2) to put up that message as part of the Import Cluster that's present all the time.
(3.3) let the user struggle and figure it out him/herself by doing the Import Cluster and letting it fail again, and hopefully, we think about showing some helpful message in the Task Details or in the message column (in the Clusters List) to say they need to do an Unmanage before an Import Cluster again.  We probably need to add something about this in the Troubleshooting section (docs) as well.

Remember, the user's goal is to import a cluster, and the fact that there were some issues that require another Import to happen but necessitates cleanup of the data does not mean we need to force the user to learn about the Tendrl internals (by having to do an Unmanage then Import to actually accomplish what the actual goal of import is).

**Comment by r0h4n** [comment]
Please comment if anything is missing, I will reopen this one
